src/native_mux_impl/stdio.rs

```diff
@@ -28,25 +29,73 @@ fn get_null_fd() -> Result<RawFd, Error> {
 }
 
 pub(crate) enum Fd {
-    PipeReadEnd(PipeRead),
-    PipeWriteEnd(PipeWrite),
-
+    Owned(OwnedFd),
     Borrowed(RawFd),
     Null,
 }
 
+fn cvt(ret: c_int) -> io::Result<c_int> {
+    if ret == -1 {
+        Err(io::Error::last_os_error())
+    } else {
+        Ok(ret)
+    }
+}
+
+fn set_blocking_inner(fd: RawFd) -> io::Result<()> {
+    let flags = cvt(unsafe { fcntl(fd, F_GETFL) })?;
+    cvt(unsafe { fcntl(fd, F_SETFL, flags & (!O_NONBLOCK)) })?;
+
+    Ok(())
+}
+
+fn set_blocking(fd: RawFd) -> Result<(), Error> {
+    set_blocking_inner(fd).map_err(Error::ChildIo)
+}
+
 impl Fd {
     pub(crate) fn as_raw_fd_or_null_fd(&self) -> Result<RawFd, Error> {
         use Fd::*;
 
         match self {
-            PipeReadEnd(fd) => Ok(AsRawFd::as_raw_fd(fd)),
-            PipeWriteEnd(fd) => Ok(AsRawFd::as_raw_fd(fd)),
-
+            Owned(owned_fd) => Ok(owned_fd.as_raw_fd()),
             Borrowed(rawfd) => Ok(*rawfd),
             Null => get_null_fd(),
         }
     }
+
+    /// # Safety
+    ///
+    /// `T::into_raw_fd` must return a valid fd and transfers
+    /// the ownershipt of it.
+    unsafe fn new_owned<T: IntoRawFd>(fd: T) -> Result<Self, Error> {
+        let raw_fd = fd.into_raw_fd();
+        Ok(Fd::Owned(OwnedFd::from_raw_fd(raw_fd)))
+    }
+}
+
+impl TryFrom<PipeRead> for Fd {
```

Comment:
I think we should make these private functions instead of implementing the trait, because otherwise `PipeRead` and `PipeWrite` become part of our public API whereas I don't think they currently are.

Reply:
Emmm they are already part of the public api.

These TryFrom impl is present in 0.9.0, but we can remove that in the next major release.

Reply:
I'm confused — they show as added in this PR? Where did they live before? If they truly exited before, we should at least mark them as deprecated.

Reply:
They indeed [exists](https://github.com/openssh-rust/openssh/blob/5282f39b71d9750cc294d145c3d82ad57634ae0f/src/stdio.rs#L86) before this PR.

I've added too many trait impl and back then I didn't think of the effect of it on the API.

I can certainly mark it as deprecated, though due to https://github.com/rust-lang/rust/issues/39935 it will be warned as "useless".

Reply:
Ah, those impls are for `Stdio`, not for `Fd`. And, looking at it some more, since `Fd` isn't a public type, this particular impl isn't a problem. Those impls are, though not much to do about those now :sweat_smile:

Reply:
Done!

I've also updated their doc to reflect that, since `#[deprecate]` does not change the doc at all.